File: frontend/src/lib/api/myersBriggs.ts
/**
 * Myers-Briggs API client
 *
 * Part of Phase 6: Myers-Briggs personality type integration
 */
import { apiClient, getErrorMessage } from './client'

// ==================== Type Definitions ====================

export interface DichotomyScore {
  dichotomy: string
  preference: string
  strength: number
  first_option: string
  second_option: string
  first_score: number
  second_score: number
  contributing_factors: string[]
}

export interface CognitiveFunction {
  function: string
  name: string
  position: string
  description: string
}

export interface AstrologicalCorrelation {
  element: string
  influences: Record<string, number>
  explanation: string
}

export interface MBTypeResponse {
  id: string | null
  birth_data_id: string
  type_code: string
  type_name: string
  temperament: string
  dichotomies: DichotomyScore[]
  preference_strengths: Record<string, number>
  description: string
  strengths: string[]
  challenges: string[]
  cognitive_stack: CognitiveFunction[] | null
  correlations: AstrologicalCorrelation[] | null
  calculation_info: Record<string, any>
  created_at: string
}

export interface MBTypeInfo {
  type_code: string
  name: string
  temperament: string
  description: string
  cognitive_functions: string[]
  famous_examples: string[]
  percentage: string
}

export interface MBTypesListResponse {
  types: MBTypeInfo[]
  count: number
}

export interface MBDichotomyInfo {
  code: string
  name: string
  first_pole: string
  second_pole: string
  first_description: string
  second_description: string
}

export interface MBDichotomiesListResponse {
  dichotomies: MBDichotomyInfo[]
  count: number
}

export interface MBTypeInterpretationResponse {
  type_code: string
  interpretation: string
}

export interface MBFullReadingResponse {
  birth_data_id: string
  type_code: string
  reading: string
  sections: Record<string, string>
  generated_at: string
}

// ==================== API Functions ====================

/**
 * Calculate Myers-Briggs type for birth data
 */
export async function calculateMBType(
  birthDataId: string,
  options: {
    include_cognitive_stack?: boolean
    include_correlations?: boolean
  } = {}
): Promise<MBTypeResponse> {
  try {
    const params = new URLSearchParams()
    if (options.include_cognitive_stack !== undefined) {
      params.append('include_cognitive_stack', String(options.include_cognitive_stack))
    }
    if (options.include_correlations !== undefined) {
      params.append('include_correlations', String(options.include_correlations))
    }

    const url = `/myers-briggs/calculate/${birthDataId}${params.toString() ? '?' + params.toString() : ''}`
    const response = await apiClient.get(url, { timeout: 30000 })
    return response.data
  } catch (error) {
    throw new Error(getErrorMessage(error))
  }
}

/**
 * Calculate MB type via POST with full options
 */
export async function calculateMBTypePost(
  birthDataId: string,
  options: {
    include_cognitive_stack?: boolean
    include_correlations?: boolean
  } = {}
): Promise<MBTypeResponse> {
  try {
    const response = await apiClient.post('/myers-briggs/calculate', {
      birth_data_id: birthDataId,
      ...options
    }, { timeout: 30000 })
    return response.data
  } catch (error) {
    throw new Error(getErrorMessage(error))
  }
}

// ==================== Reference Data ====================

/**
 * Get list of all 16 MB types
 */
export async function getTypesList(): Promise<MBTypesListResponse> {
  try {
    const response = await apiClient.get('/myers-briggs/types')
    return response.data
  } catch (error) {
    throw new Error(getErrorMessage(error))
  }
}

/**
 * Get info for a specific MB type
 */
export async function getTypeInfo(typeCode: string): Promise<MBTypeInfo> {
  try {
    const response = await apiClient.get(`/myers-briggs/types/${typeCode}`)
    return response.data
  } catch (error) {
    throw new Error(getErrorMessage(error))
  }
}

/**
 * Get list of all 4 dichotomies
 */
export async function getDichotomiesList(): Promise<MBDichotomiesListResponse> {
  try {
    const response = await apiClient.get('/myers-briggs/dichotomies')
    return response.data
  } catch (error) {
    throw new Error(getErrorMessage(error))
  }
}

// ==================== AI Interpretations ====================

/**
 * Get AI interpretation for MB type
 */
export async function getTypeInterpretation(
  typeCode: string,
  temperament: string,
  strengths?: string[],
  challenges?: string[]
): Promise<MBTypeInterpretationResponse> {
  try {
    const response = await apiClient.post('/myers-briggs/ai/interpret-type', {
      type_code: typeCode,
      temperament,
      strengths: strengths || [],
      challenges: challenges || []
    }, { timeout: 60000 })
    return response.data
  } catch (error) {
    throw new Error(getErrorMessage(error))
  }
}

/**
 * Get full AI reading for MB type
 */
export async function getFullMBReading(
  birthDataId: string,
  includeCorrelations: boolean = false
): Promise<MBFullReadingResponse> {
  try {
    const response = await apiClient.post('/myers-briggs/ai/full-reading', {
      birth_data_id: birthDataId,
      include_correlations: includeCorrelations
    }, { timeout: 60000 })
    return response.data
  } catch (error) {
    throw new Error(getErrorMessage(error))
  }
}

// ==================== Helper Functions ====================

/**
 * Get display name for MB type
 */
export function getTypeDisplayName(typeCode: string): string {
  const names: Record<string, string> = {
    'ISTJ': 'The Inspector',
    'ISFJ': 'The Protector',
    'INFJ': 'The Advocate',
    'INTJ': 'The Architect',
    'ISTP': 'The Virtuoso',
    'ISFP': 'The Adventurer',
    'INFP': 'The Mediator',
    'INTP': 'The Logician',
    'ESTP': 'The Entrepreneur',
    'ESFP': 'The Entertainer',
    'ENFP': 'The Campaigner',
    'ENTP': 'The Debater',
    'ESTJ': 'The Executive',
    'ESFJ': 'The Consul',
    'ENFJ': 'The Protagonist',
    'ENTJ': 'The Commander',
  }
  return names[typeCode.toUpperCase()] || typeCode
}

/**
 * Get color for MB type (by temperament)
 */
export function getTypeColor(typeCode: string): string {
  const code = typeCode.toUpperCase()

  // Colors by temperament
  if (code.includes('SJ')) {  // Guardian
    return '#3B82F6' // Blue
  } else if (code.includes('SP')) {  // Artisan
    return '#F59E0B' // Amber
  } else if (code.includes('NF')) {  // Idealist
    return '#10B981' // Emerald
  } else if (code.includes('NT')) {  // Rational
    return '#8B5CF6' // Purple
  }

  return '#6B7280' // Gray fallback
}

/**
 * Get temperament name from type code
 */
export function getTemperamentName(typeCode: string): string {
  const code = typeCode.toUpperCase()

  if (code[1] === 'S' && code[3] === 'J') return 'Guardian'
  if (code[1] === 'S' && code[3] === 'P') return 'Artisan'
  if (code[1] === 'N' && code[2] === 'F') return 'Idealist'
  if (code[1] === 'N' && code[2] === 'T') return 'Rational'

  return 'Unknown'
}

/**
 * Get temperament color
 */
export function getTemperamentColor(temperament: string): string {
  const colors: Record<string, string> = {
    'Guardian': '#3B82F6',
    'Artisan': '#F59E0B',
    'Idealist': '#10B981',
    'Rational': '#8B5CF6',
  }
  return colors[temperament] || '#6B7280'
}

/**
 * Get dichotomy display name
 */
export function getDichotomyDisplayName(code: string): string {
  const names: Record<string, string> = {
    'E/I': 'Extraversion / Introversion',
    'S/N': 'Sensing / Intuition',
    'T/F': 'Thinking / Feeling',
    'J/P': 'Judging / Perceiving',
  }
  return names[code] || code
}

/**
 * Get preference description
 */
export function getPreferenceDescription(letter: string): string {
  const descriptions: Record<string, string> = {
    'E': 'Extraversion - Energized by interaction',
    'I': 'Introversion - Energized by reflection',
    'S': 'Sensing - Focus on concrete facts',
    'N': 'Intuition - Focus on patterns and possibilities',
    'T': 'Thinking - Decisions based on logic',
    'F': 'Feeling - Decisions based on values',
    'J': 'Judging - Prefer structure and closure',
    'P': 'Perceiving - Prefer flexibility and options',
  }
  return descriptions[letter.toUpperCase()] || letter
}

/**
 * Get cognitive function display name
 */
export function getCognitiveFunctionDisplayName(code: string): string {
  const names: Record<string, string> = {
    'Se': 'Extraverted Sensing',
    'Si': 'Introverted Sensing',
    'Ne': 'Extraverted Intuition',
    'Ni': 'Introverted Intuition',
    'Te': 'Extraverted Thinking',
    'Ti': 'Introverted Thinking',
    'Fe': 'Extraverted Feeling',
    'Fi': 'Introverted Feeling',
  }
  return names[code] || code
}

/**
 * Get function position color
 */
export function getFunctionPositionColor(position: string): string {
  const colors: Record<string, string> = {
    'Dominant': '#10B981',   // Green
    'Auxiliary': '#3B82F6',  // Blue
    'Tertiary': '#F59E0B',   // Amber
    'Inferior': '#EF4444',   // Red
  }
  return colors[position] || '#6B7280'
}

/**
 * Format preference strength as percentage
 */
export function formatPreferenceStrength(score: number): string {
  return `${Math.round(score)}%`
}

/**
 * Get letter pair from dichotomy code
 */
export function getDichotomyLetters(dichotomy: string): [string, string] {
  const pairs: Record<string, [string, string]> = {
    'E/I': ['E', 'I'],
    'S/N': ['S', 'N'],
    'T/F': ['T', 'F'],
    'J/P': ['J', 'P'],
  }
  return pairs[dichotomy] || ['?', '?']
}

/**
 * Check if type is introverted
 */
export function isIntroverted(typeCode: string): boolean {
  return typeCode.toUpperCase().startsWith('I')
}

/**
 * Check if type is intuitive
 */
export function isIntuitive(typeCode: string): boolean {
  return typeCode.toUpperCase()[1] === 'N'
}

/**
 * Check if type is thinking
 */
export function isThinking(typeCode: string): boolean {
  return typeCode.toUpperCase()[2] === 'T'
}

/**
 * Check if type is judging
 */
export function isJudging(typeCode: string): boolean {
  return typeCode.toUpperCase().endsWith('J')
}

File: backend/app/schemas/myers_briggs.py
"""
Myers-Briggs Schemas

Pydantic schemas for Myers-Briggs personality type calculations and responses.
Based on astrological correspondences to derive personality type.
"""

from pydantic import BaseModel, Field
from typing import Dict, List, Optional, Any
from datetime import datetime
from enum import Enum


# ==============================================================================
# ENUMS
# ==============================================================================

class MBDichotomy(str, Enum):
    """The four dichotomies of Myers-Briggs."""
    EXTRAVERSION_INTROVERSION = "E/I"
    SENSING_INTUITION = "S/N"
    THINKING_FEELING = "T/F"
    JUDGING_PERCEIVING = "J/P"


class MBPreference(str, Enum):
    """Individual preference codes."""
    EXTRAVERSION = "E"
    INTROVERSION = "I"
    SENSING = "S"
    INTUITION = "N"
    THINKING = "T"
    FEELING = "F"
    JUDGING = "J"
    PERCEIVING = "P"


class MBType(str, Enum):
    """The 16 Myers-Briggs personality types."""
    ISTJ = "ISTJ"
    ISFJ = "ISFJ"
    INFJ = "INFJ"
    INTJ = "INTJ"
    ISTP = "ISTP"
    ISFP = "ISFP"
    INFP = "INFP"
    INTP = "INTP"
    ESTP = "ESTP"
    ESFP = "ESFP"
    ENFP = "ENFP"
    ENTP = "ENTP"
    ESTJ = "ESTJ"
    ESFJ = "ESFJ"
    ENFJ = "ENFJ"
    ENTJ = "ENTJ"


class MBTemperament(str, Enum):
    """Keirsey's four temperaments."""
    GUARDIAN = "Guardian"      # SJ types
    ARTISAN = "Artisan"        # SP types
    IDEALIST = "Idealist"      # NF types
    RATIONAL = "Rational"      # NT types


class CognitiveFunctionPosition(str, Enum):
    """Cognitive function stack positions."""
    DOMINANT = "Dominant"
    AUXILIARY = "Auxiliary"
    TERTIARY = "Tertiary"
    INFERIOR = "Inferior"


# ==============================================================================
# NESTED RESPONSE MODELS
# ==============================================================================

class DichotomyScore(BaseModel):
    """Score for a single dichotomy."""
    dichotomy: str = Field(..., description="The dichotomy (E/I, S/N, T/F, J/P)")
    preference: str = Field(..., description="The dominant preference letter")
    strength: float = Field(..., ge=0, le=100, description="Strength of preference (0-100)")
    first_option: str = Field(..., description="First option (E, S, T, J)")
    second_option: str = Field(..., description="Second option (I, N, F, P)")
    first_score: float = Field(..., description="Score for first option")
    second_score: float = Field(..., description="Score for second option")
    contributing_factors: List[str] = Field(
        default_factory=list,
        description="Astrological factors contributing to this score"
    )

    class Config:
        json_schema_extra = {
            "example": {
                "dichotomy": "E/I",
                "preference": "I",
                "strength": 72.5,
                "first_option": "E",
                "second_option": "I",
                "first_score": 27.5,
                "second_score": 72.5,
                "contributing_factors": ["Sun in Cancer", "Moon in Pisces", "Ascendant in Virgo"]
            }
        }


class CognitiveFunction(BaseModel):
    """A cognitive function in the stack."""
    function: str = Field(..., description="Function code (e.g., Fi, Te, Ni)")
    name: str = Field(..., description="Full name (e.g., Introverted Feeling)")
    position: CognitiveFunctionPosition = Field(..., description="Position in stack")
    description: str = Field(..., description="What this function does")


class TypeProfile(BaseModel):
    """Profile information for an MBTI type."""
    type_code: str = Field(..., description="4-letter type code")
    name: str = Field(..., description="Type nickname (e.g., The Architect)")
    temperament: MBTemperament = Field(..., description="Keirsey temperament")
    description: str = Field(..., description="Type description")
    strengths: List[str] = Field(..., description="Key strengths")
    challenges: List[str] = Field(..., description="Potential challenges")
    cognitive_stack: List[CognitiveFunction] = Field(
        ..., description="Cognitive function stack (dominant to inferior)"
    )


class AstrologicalCorrelation(BaseModel):
    """How a birth chart element correlates to MB preferences."""
    element: str = Field(..., description="Astrological element (e.g., Sun in Aries)")
    influences: Dict[str, float] = Field(
        ..., description="Influence weights on each preference"
    )
    explanation: str = Field(..., description="Why this correlation exists")


# ==============================================================================
# REQUEST SCHEMAS
# ==============================================================================

class MBCalculationRequest(BaseModel):
    """Request to calculate Myers-Briggs type from birth data."""
    birth_data_id: str = Field(..., description="UUID of the birth data record")

    include_cognitive_stack: bool = Field(
        default=True,
        description="Whether to include cognitive function stack"
    )
    include_correlations: bool = Field(
        default=False,
        description="Whether to include astrological correlation details"
    )

    class Config:
        json_schema_extra = {
            "example": {
                "birth_data_id": "123e4567-e89b-12d3-a456-426614174000",
                "include_cognitive_stack": True,
                "include_correlations": False
            }
        }


class MBInterpretationRequest(BaseModel):
    """Request for AI interpretation of MB type."""
    type_code: str = Field(..., description="4-letter type code")
    context: Optional[Dict[str, Any]] = Field(
        default=None,
        description="Additional context for interpretation"
    )


# ==============================================================================
# RESPONSE SCHEMAS
# ==============================================================================

class MBTypeResponse(BaseModel):
    """Complete Myers-Briggs type response."""
    id: Optional[str] = Field(None, description="Calculation ID if saved")
    birth_data_id: str = Field(..., description="Birth data ID")

    # Core type
    type_code: str = Field(..., description="4-letter type code (e.g., INTJ)")
    type_name: str = Field(..., description="Type nickname")
    temperament: str = Field(..., description="Keirsey temperament")

    # Dichotomy scores
    dichotomies: List[DichotomyScore] = Field(
        ..., description="Scores for each of the 4 dichotomies"
    )

    # Preference strengths summary
    preference_strengths: Dict[str, float] = Field(
        ..., description="Map of each letter to its strength (0-100)"
    )

    # Type description
    description: str = Field(..., description="Description of this type")
    strengths: List[str] = Field(..., description="Key strengths")
    challenges: List[str] = Field(..., description="Potential challenges")

    # Cognitive functions
    cognitive_stack: Optional[List[CognitiveFunction]] = Field(
        None, description="Cognitive function stack if requested"
    )

    # Astrological correlations
    correlations: Optional[List[AstrologicalCorrelation]] = Field(
        None, description="Astrological correlations if requested"
    )

    # Calculation metadata
    calculation_info: Dict[str, Any] = Field(
        default_factory=dict,
        description="Calculation metadata"
    )

    created_at: Optional[datetime] = Field(None, description="When calculated")

    class Config:
        json_schema_extra = {
            "example": {
                "type_code": "INTJ",
                "type_name": "The Architect",
                "temperament": "Rational",
                "dichotomies": [
                    {"dichotomy": "E/I", "preference": "I", "strength": 72.5}
                ],
                "description": "Strategic, logical, and independent thinkers"
            }
        }


class MBTypeInfo(BaseModel):
    """Information about a single MB type."""
    type_code: str
    name: str
    temperament: str
    description: str
    cognitive_functions: List[str]
    famous_examples: List[str] = Field(default_factory=list)
    percentage: str = Field(..., description="Estimated population percentage")


class MBTypesListResponse(BaseModel):
    """Response for listing all 16 types."""
    types: List[MBTypeInfo]
    count: int = 16


class MBDichotomyInfo(BaseModel):
    """Information about a dichotomy."""
    code: str
    name: str
    first_pole: str
    second_pole: str
    first_description: str
    second_description: str


class MBDichotomiesListResponse(BaseModel):
    """Response for listing all dichotomies."""
    dichotomies: List[MBDichotomyInfo]
    count: int = 4


# ==============================================================================
# AI INTERPRETATION RESPONSES
# ==============================================================================

class MBTypeInterpretationResponse(BaseModel):
    """AI interpretation of MB type."""
    type_code: str
    interpretation: str


class MBFullReadingResponse(BaseModel):
    """AI-generated full MB reading."""
    birth_data_id: str
    type_code: str
    reading: str
    sections: Dict[str, str] = Field(
        default_factory=dict,
        description="Reading sections (overview, strengths, challenges, growth, relationships)"
    )
    generated_at: datetime = Field(default_factory=datetime.utcnow)
